Supported popup menu window

jkim-julie opened the pull request:
It included five commits. Supporting Popup Menu basically required '
Fix Showing/Hiding, Placing and Events for Menu Windows (#73)' from previous branch but it had dependency on '
First attempt to remove decorations from small windows'.
1) Getting '
First attempt to remove decorations from small windows' without changes from services
2) Applying 'Showing/Hiding, Placing and Events for Menu Windows (#73)''
3) Making it work on new wayland port without mus.
4) Applyng 'Add popup window type support' except services.
5) Making it work on new wayland port without mus.

We could combine '2) & 3)' and '4) & 5)' but I left them separately for a review.
Once you finish review, I could squash commits. WDYT?

With this branch, I confirmed drop-down list and menu from 3-dot work fine.

msisov committed:
Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

This adaptation makes it possible to show/hide and place menu windows
properly in regards to their parent surface. What is more, this patch
fixes event locations sent to the client from wayland.
Conflicts:
	ui/ozone/platform/wayland/wayland_window.cc

fixup! [Wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

This resolves crash when a user requests to open a submenu window

fixup! [Wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

Typo fix

fixup! [Wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

Fix crash when menus are destroyed, but aura is still trying to hide them.
It may happen when wayland decides to destroy the menu (wrong serial number, for example).
The problem here is that we set a child window pointer to the parent, but never remove it.

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows

Ensure that a pointer to child window is invalidated as soon as xdg_popup_ is
destroyed. This fixes a random crash, when the wrong pointer to a child
deletes not the top most popup window, which results in a dismiss signal from
wayland and all windows are being closed.

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

This patch makes opening submenus on mouse hover to work. Basically, creating
popups requires serials, which correspond to grab events (those
are touch, key press or button press events). As long as other
events' serials are not used by us, don't save such serials that
correspond to Enter/Leave and etc.

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows

Remove setting serial from WaylandKeyboard::Modifiers() as well.

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows

Release buffer as soon as the popups are destroyed.

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

Add nullcheck

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

Minor: No need to use ::get to null-check unique_ptr instances.

Issue #266

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

Pass parent WaylandWindow instead of wl_surface to avoid searching for a
WaylandWindow by wl_surface. This is an effort to prepare upstreaming
the local changes.

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

Driven by base::MakeUnique -> std::make_unique.

TBR=msisov@igalia.com

Issue #266

fixup! [ozone/wayland] Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

fixup! Fix Showing/Hiding, Placing and Events for Menu Windows (#73)

It delivered client::WindowType to CreateAndSetDefaultPlatformWindow
to let ui::PlatformWindowDelegate know window type because
wayland_window needs it directly  on Initialize().
It also defered creating xdg popup when it doesn't have bounds. It
will try to create xdg popup again on showing.

It introduced WindowTreeHostPlatform::InitProperties to keep
properties from Widget::InitParams.

Issue #417

msisov committed:
Add popup window type support.

This commit adds a popup window type to ozone windows.
Popup windows are drop down windows, for example.

Issue #417

fixup! Add popup window type support.

It added case handling on WindowTreeHostPlatform::GetWindowType().

jkim-julie committed:
First attempt to remove decorations from small windows

This patch sets XSetWindowAttributes::override_redirect to
True, which removes different properties from windows.
At this stage, all windows with width() less than 600 are treated
as menus and override_redirect is set to True.

It's almost same as we had with mus except services.

Issue #38
Issue #417

msisov commented on ui/ozone/platform/wayland/wayland_window.cc:
```diff
@@ -109,11 +109,22 @@ bool WaylandWindow::Initialize() {
+      CreateXdgSurface();
+      break;
+    default:
+      NOTREACHED() << "Not supported window type: type=" << ui_window_type;
```
This makes the browser crashing immediately, when it tries to create tooltip windows, for example. 

We had this NOTREACHED() when tooltips were disabled on the aura level. 

Can you please make another [TEMP] patch, which unites the
case ui::PlatformWindowType::PLATFORM_WINDOW_TYPE_WINDOW:
and default cases and calls CreateXdgSurface instead?

`case ui::PlatformWindowType::PLATFORM_WINDOW_TYPE_WINDOW:`
 `default:`
`   CreateXdgSurface();`
 `  break;`

jkim-julie replied:
Sorry, I realized that crash just before. I updated branch with removing converting tooltip window type at  GetWindowType() minutes ago because there is no tooltip handling yet. WDYT?

msisov replied:
add todo there.

msisov commented on ui/aura/window_tree_host_platform.h:
```diff
@@ -56,7 +57,7 @@ class AURA_EXPORT WindowTreeHostPlatform : public WindowTreeHost,
   // Creates a ui::PlatformWindow appropriate for the current platform and
   // installs it at as the PlatformWindow for this WindowTreeHostPlatform.
-  void CreateAndSetDefaultPlatformWindow();
+  void CreateAndSetDefaultPlatformWindow(const client::WindowType type);
```
What if we add a method called CreateAndSetPlatformWindowWithParameters(const Widget::InitParams& params), which will store Widget::InitParams, and then call CreateAndSetDefaultPlatformWindow. 

Then, PlatformWindowDelegate can declare getters we need such as GetWindowType, GetParentWindowWidget (this will be used instead of WaylandConnection::GetCurrentFocusedWindow) and etc?

jkim-julie replied:
OK. I'll update.

msisov commented on ui/ozone/platform/wayland/wayland_window.cc:
```diff
@@ -96,10 +123,33 @@ bool WaylandWindow::Initialize() {
 
+void WaylandWindow::CreateXdgPopup() {
+  if (!parent_window_)
+    parent_window_ = connection_->GetCurrentFocusedWindow();
```
Check comment at https://github.com/Igalia/chromium/pull/428/commits/1807c9636035bc23c34d5a3176199cf68ec4b44b#diff-ab710017f2cbbeef686bc6cc4c5a338cR60

msisov replied:
Looks like you didn't pay attention to this.

Would you please check the following commit - 4de6b42a0ee104503cfb2763a006784f1812b836 ?

Afaik, Widget::InitParams has a  gfx::NativeView parent, which you can reuse to extract a gfx::AcceleratedWidget from it. Then, you could call delegate_->GetParentWindow(gfx::AcceleratedWidget* parent_widget) and take it instead. Once you have it, just call WaylandConnection::GetWindow(widget) to get a parent WaylandWindow.

msisov replied:
Looks like you didn't pay attention to this.

Would you please check the following commit - 4de6b42 ?

Afaik, Widget::InitParams has a gfx::NativeView parent, which you can reuse to extract a gfx::AcceleratedWidget from it. Then, you could call delegate_->GetParentWindow(gfx::AcceleratedWidget* parent_widget) and take it instead. Once you have it, just call WaylandConnection::GetWindow(widget) to get a parent WaylandWindow.

jkim-julie replied:
@msisov, please look into Issue #424. I created issue for parent-child relationship based on pads we listed up tasks and it's prepared at my local in order to send another PR once this PR is merged. I thought we'd better keep previous history. Do you think we have to handle it with this branch together?

msisov replied:
I'd better want you to do it right way here rather than duplicating our efforts. It will ease upstreaming efforts. I'm 100% sure that the way how current patch does won't be accepted.

msisov commented:
Yes, please squash those.

jkim-julie commented:
I added 'TODD' at WindowTreeHostPlatform::GetWindowType.

I thought we'd better to keep necessary properties, not all of them from Widget::InitParams, store them at aura area because they are used at aura and convert them at DesktopWindowTreeHostPlatform before delivering params.
I declared new struct WindowTreeHostPlatform::InitProperties and created DesktopWindowTreeHostPlatform::ConvertParamsToProperties() to get properties required on creating window from Widget::InitParams. The properties are stored at WindowTreeHostPlatform::properties_ as the type, InitProperties. When we need to add more properties from Widget::InitParams, we can add it at DesktopWindowTreeHostPlatform::ConvertParamsToProperties. I left 'TODO' as well.

I squashed commits.
PTAL, Thanks.

msisov commented on ui/aura/window_tree_host_platform.h:
```diff
@@ -57,6 +65,8 @@ class AURA_EXPORT WindowTreeHostPlatform : public WindowTreeHost,
   // Creates a ui::PlatformWindow appropriate for the current platform and
   // installs it at as the PlatformWindow for this WindowTreeHostPlatform.
   void CreateAndSetDefaultPlatformWindow();
+  void CreateAndSetDefaultPlatformWindowWithProperties(
```
s/CreateAndSetDefaultPlatformWindowWithProperties/CreateAndSetPlatformWindowWithProperties

jkim-julie commented:
It included #424 and commit 4de6b42a0ee104503cfb2763a006784f1812b836 at https://github.com/Igalia/chromium/pull/428/commits/c38cd3d8eb0b984e3c3da292b3a458c29dd23fd7.

msisov commented on ui/aura/window_tree_host_platform.h:
```diff
@@ -29,6 +30,15 @@ class WindowPort;
                                            public ui::PlatformWindowDelegate {
  public:
+  struct AURA_EXPORT InitProperties {
+    InitProperties();
```
Afaik, chromium style needs a dtor as well.

jkim-julie replied:
Done.
